### DISCUSSION TITLE
correct procedure for manual installation of features

### DISCUSSION THREAD

**Discussion by gireeshpunathil** [eclipse-equinox/p2#456] [General]
In scenarios where there is a headless pipeline driving feature installations, we cannot use the IDE's help -> install s/w or `p2.director` application. So some of my users resort to:
 - manually copy the features and plugins to their respective folders
 - and then update `bundles.info` file and add entries of the new comers

questions:
 - are these steps fool proof / safe?
 - are these steps sufficient?

**Comment by laeubi** [comment]
> we cannot use the [...] p2.director application.

Why?

> are these steps fool proof / safe?

No

> are these steps sufficient?

No

You should really use the [director application](https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/guide/p2_director.html?cp=2_0_20_2) to install things from command-line:

- https://tycho.eclipseprojects.io/doc/latest/tycho-p2-director-plugin/director-mojo.html
- https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/guide/p2_director.html

**Reply by merks** [reply]
Maybe dropins still work:

https://wiki.eclipse.org/Equinox/p2/Getting_Started#Dropins

**Reply by laeubi** [reply]
> Maybe dropins still work:
> 
> https://wiki.eclipse.org/Equinox/p2/Getting_Started#Dropins

Will this execute touch point actions? Any checks done that the install will work? Detect conflicting dependencies?

**Reply by gireeshpunathil** [reply]
@merks - does droppins mandate a running system? AFAIK, in pipeline mode, a running system will be missing

**Reply by laeubi** [reply]
> @merks - does droppins mandate a running system? AFAIK, in pipeline mode, a running system will be missing

How do "manually copy the features and plugins" and " update bundles.info file and add entries of the new comers" work if there is nothing running? there must be something running ;-)
Or is it [Sneakernet](https://en.wikipedia.org/wiki/Sneakernet)?

**Reply by gireeshpunathil** [reply]
@laeubi - they can use `p2.director`, but looks like its running time is in the order of seconds, if not minutes. If we precisely understand the final effect of `p2.director` on the system, performing those same sequences using a script would save time - this is the premise in which the user is seeing the case.

**Reply by gireeshpunathil** [reply]
> there must be something running ;-)

running on a bare metal with OS commands and custom scripts (awk/sed etc.) that can perform fs operations.

**Reply by laeubi** [reply]
Tycho has some speed optimization for director, still in most cases its better to be slower but correct than faster but failing, how many builds per minute we are talking about that it maters if it takes a few seconds longer?

**Reply by merks** [reply]
Precisely trying to replicate what p2 is doing in a script sounds error broken at best.   As I understand it, the dropins folder should be reconciled properly on startup.  But I wouldn't bet my life on it working well.  The p2 director knows what it's doing...

**Reply by laeubi** [reply]
> > there must be something running ;-)
> 
> running on a bare metal with OS commands and custom scripts (awk/sed etc.) that can perform fs operations.

But how does the bundles / feature / product come from then?!?

**Reply by gireeshpunathil** [reply]
I don't know, but I would assume that they fetch bundles from some private curated repo time to time (or when a new version is available) and then ignite the pipeline to install things and redistribute the new builds..

**Comment by gireeshpunathil** [comment]
thanks @laeubi and @merks for the assertions: especially on the resilience and the performance part. I now understand that `p2.director` is the right and recommended way to install things in the command line, and short-circuiting that is not reliable. I will keep this in my best practice list.

**Reply by merks** [reply]
One would generally expect that a simple automated check for updates would suffice.  That is the intended purpose...

**Reply by laeubi** [reply]
Yes p2 director is somehow "guaranteed" to work as it is used inside Eclipse and Tycho, everything else might work (under some special circumstances) but might fail anytime without notice.

Beside that optimizing that parts will help to improve it for everyone and in the future, while local solution tend to [bit-rot](https://en.wikipedia.org/wiki/Software_rot) over time and prevents from further adopting of new versions because "something" suddenly fails.

**Reply by gireeshpunathil** [reply]
I just double checked with the user, they said that they did not have `p2.director` earlier, and as they cannot `run` eclipse IDE (they package and redistribute products to their customers, so they don't have a dev env / platforms etc.) to check for updates or install s/w through first class menu options, manual placement was what they used. In such circumstances dropins may not be ideal, because if the dropped bundles have dependency issues, that will show up in their end customer's environment, which is too late.

anyways, I understand your reasons and agree with that, thank you!

**Reply by laeubi** [reply]
If the package a product why they need to modify it AFTERWARDS?
Also note that the Tycho mojo can run without any Eclipse installation and install things into any product that is P2 managed.

**Comment by mickaelistria** [comment]
Basically, p2 director is the very only way to do this reliably (either directly, or involving Tycho or Oomph or whatever). Trying to do things reliably without p2 director will require you to more or less reimplement p2 director with awk or other, to get proper deps resolution, touchpoint execution and so on.
p2 director can be package as a complete application, with Java embedded and so on. If packaged well, it should be as easy to use it as any other shell command.

**Reply by laeubi** [reply]
> p2 director can be package as a complete application, with Java embedded and so on. If packaged well, it should be as easy to use it as any other shell command.

@mickaelistria should we probably provide an "p2" product as an EPP like the installer? This then could even use a minimal jre so would be quite interesting for people not wanting to either build their self or use a full blown sdk...

**Reply by merks** [reply]
FYI, the Eclipse Installer includes the director application:

![image](https://github.com/eclipse-equinox/p2/assets/208716/b12910e2-32da-4d13-b982-d0d5b77c59e2)

It also has a minimal JRE, so interesting options are already available...

**Reply by laeubi** [reply]
> FYI, the Eclipse Installer includes the director application:
> It also has a minimal JRE, so interesting options are already available...

If just no one knows its more a hidden feature :-)

I suppose that installer also includes oomph, so a dedicated p2 application could even be more minimal... It would even be cool to have it include into the usual package managers so one can just do `p2 <whatever>` on any linux distribution (eclipse.installer of course as well).
